pep-XXXX.rst

```diff
@@ -0,0 +1,782 @@
+``typing.Concatenate`` to help describe functions that pass ``*args``
+and ``**kwargs`` to callbacks, which is very common with
+decorators. This made ``typing.Callable`` more powerful but also more
+complicated.
+
+Problems with ``typing.Callable``
+---------------------------------
+
+Empirically `we have found
+<https://github.com/pradeep90/annotation_collector#typed-projects---callable-type>`_
+that it is common for library authors to make use of untyped or
+partially-typed callables (e.g. ``Callable[..., Any]`` or a bare
+``Callable``) which we believe is partially a result of the existing
+types being hard to use. But using this kind of partially-typed
+callable can negate the benefits of static typing. For example,
+consider the following code::
+
+    from typing import Any, Callable
+
+    def with_retries(f: Callable[..., Any]) -> Callable[..., Any]:
+        def wrapper(retry_once, *args, **kwargs):
+            if retry_once:
+                try: return f(*args, **kwargs)
+                except Exception: pass
+            return f(*args, **kwargs)
+        return wrapper
+
+    @with_retries
+    def f(x: int) -> int:
+        return x
+
+
+    f(True, z=15)
+
+The decorator above is clearly not intended to modify the type of the
+function it wraps, but because it uses ``Callable[..., Any]`` it
+accidentally eliminates the annotations on ``f``, and type checkers
+will accept the code above even though it is sure to crash at
+runtime. A correct version of this code would look like this::
+
+    from typing import Any, Callable, Concatenate, ParamSpec, TypeVar
+
+    V = TypeVar("R")
+    P = ParamSpec("P")
+
+    def with_retries(f: Callable[P, R]) -> Callable[Concatenate[bool, P] R]:
+        def wrapper(retry_once: bool, *args: P.args, **kwargs: P.kwargs):
+            ...
+        return wrapper
+
+    ...
+
+With these changes, the incorrect attempt to pass ``z`` to ``f``
+produces a typecheck error as we would like.
+
+Four usability problems with the way ``typing.Callable`` is
+represented may explain why library authors often do not use its full
+power:
+
+- It is verbose, particularly for more complex function signatures.
+- It does not visually represent the way function headers are written,
+  which can make it harder to learn and use.
+- It requires an explicit import, something we no longer require for
+  most of the other very common types after PEP 604 (``|`` for
+  ``Union`` types) and PEP 585 (generic collections)
+- It relies on two levels of nested square brackets. This can be quite
+  hard to read, especially when the function arguments themselves have
+  square brackets.
+
+With our proposed syntax, the properly-typed decorator example becomes
+concise and the type representations are visually descriptive::
+
+    from typing import Any, ParamSpec, TypeVar
+
+    V = TypeVar("R")
+    P = ParamSpec("P")
+
+    def with_retries(f: (**P) -> R) -> (bool, **P) -> R:
```

Comment:
I'm generally extremely supportive of this PEP, but I worry about this example. The number of arrows in one signature makes my eyes glaze over. The thing is, I'm 90% sure that this will be vastly improved by appropriate syntax highlighting in an IDE — but maybe that should be mentioned?

Reply:
Also simple whitespace trick might make it more readable:
```
def with_retries(
    f: (**P) -> R,
) -> (bool, **P) -> R:
```

Reply:
Can PEP document source use the `.. code-block:: python` RST directive to make the examples use the Python syntax highlighting?

Reply:
> Can PEP document source use the `.. code-block:: python` RST directive to make the examples use the Python syntax highlighting?

Not currently, I don't think, sadly — though there's a proposal to rework the way they're rendered that would give us syntax highlighting 🙏 (though I'm not sure we'd gain much from that with this PEP, anyway, since we're proposing new syntax?) https://discuss.python.org/t/request-for-comment-making-pep-rendering-self-contained/10774

Reply:
It's probably not the best intro example but my creativity gave out and the easiest realistic use case I could think of was a decorator.

We'll see if @pradeep90 can come up with a better one. The problem is that we're supposed to be highlighting the kind of situation where people give up and use partially-typed functions, which means the example has to be a little uglier than the basic intro-to-callable-type.